pkg/telemetry/telemetry.go

```diff
@@ -0,0 +1,266 @@
+		defer collect.Stop()
+
+		for {
+			select {
+			case <-t.stop:
+				close(t.stop)
+				return
+			case <-collect.C:
+			}
+			var (
+				err             error
+				underlyingValue float64
+				newStats        telemetryStats
+			)
+			t.metrics.Range(func(statName, metric interface{}) bool {
+				columnName := statName.(string)
+				promMetric := metric.(prometheus.Metric)
+				underlyingValue, err = extractMetricValue(promMetric)
+				if err != nil {
+					err = fmt.Errorf("extracting metric value of stat '%s': %w", columnName, err)
+					return false
+				}
+
+				switch columnName {
+				case "promscale_ingested_samples_total":
+					newStats.samplesIngested = underlyingValue
+				case "promscale_metrics_queries_failed_total":
+					newStats.promqlQueriesFailed = underlyingValue
+				case "promscale_metrics_queries_executed_total":
+					newStats.promqlQueriesExecuted = underlyingValue
+				case "promscale_metrics_queries_timedout_total":
+					newStats.promqlQueriesTimedout = underlyingValue
+				case "promscale_trace_query_requests_executed_total":
+					newStats.traceQueriesExecuted = underlyingValue
+				case "promscale_trace_dependency_requests_executed_total":
+					newStats.traceDependencyQueriesExecuted = underlyingValue
+				default:
+					log.Error("msg", fmt.Sprintf("unregistered column name: %s", columnName))
+				}
+
+				return true
+			})
+			if err != nil {
+				log.Debug("msg", "error extracting underlying value of a stats metric", "error", err.Error())
+				continue
+			}
+			if err := syncInfoTable(t.conn, t.uuid, newStats); err != nil {
+				log.Debug("msg", "syncing new stats", "error", err.Error())
+			}
+		}
+	}()
+}
+
+func extractMetricValue(metric prometheus.Metric) (float64, error) {
+	var internal io_prometheus_client.Metric
+	if err := metric.Write(&internal); err != nil {
+		return 0, fmt.Errorf("error writing metric: %w", err)
+	}
+	if internal.Gauge != nil {
+		return internal.Gauge.GetValue(), nil
+	} else if internal.Counter != nil {
+		return internal.Counter.GetValue(), nil
+	}
+	return 0, fmt.Errorf("both Gauge and Counter are nil")
+}
+
+// RegisterMetric registers a counter or gauge metric for telemetry purpose.
+func (t *engine) RegisterMetric(columnName string, metric prometheus.Metric) error {
+	if !isCounterOrGauge(metric) {
+		return fmt.Errorf("metric not a counter or gauge")
+	}
+	t.metrics.Store(columnName, metric)
+	return nil
+}
+
+func isCounterOrGauge(metric prometheus.Metric) bool {
+	switch metric.(type) {
+	case prometheus.Counter, prometheus.Gauge:
+		return true
+	default:
+		return false
+	}
+}
+
+// syncInfoTable stats with promscale_instance_information table.
+func syncInfoTable(conn pgxconn.PgxConn, uuid [16]byte, s telemetryStats) error {
+	lastUpdated := time.Now()
+
+	pgUUID := new(pgtype.UUID)
+	if err := pgUUID.Set(uuid); err != nil {
+		return fmt.Errorf("setting pg-uuid: %w", err)
+	}
+
+	query := fmt.Sprintf(`INSERT INTO %s.promscale_instance_information
```

Comment:
if we had a map here instead of hard-coded columns we could build this query using the keys in the map 
```
//keys is a []string of map keys
cols := "uuid, "
cols = strings.Join(keys, ",");
indexes := "$1"
lastIndex := 1
updateStatements = make([]string, 0)
for _, col := range keys {
     lastIndex += 1
     indexes += fmt.Stringf(", $%d", lastIndex)
     updateStatements = append(updateStatements, fmt.Sprintf("%s = %d", col, lastIndex) 
}


query := fmt.Sprintf(`INSERT INTO %s.promscale_instance_information(%s)
VALUES(%s)
ON CONFLICT (uuid) DO
    UPDATE SET
         %s`, schema.PSCatalog, cols, indexes, strings.Join(updateStatements, ", "))

```

Reply:
Ah, this way. An even easier would be this

```golang
columnNames := []string{"uuid", "last_updated"}
	columnValues := []interface{}{pgUUID, lastUpdated}
	indexes := []string{"$1", "$2"}
	updateStatements := []string{"last_updated = $2"}

	index := 2 // Since 1 & 2 are uuid & last_updated.
	for k, v := range stats {
		index++
		indexes = append(indexes, fmt.Sprintf("$%d", index))
		columnNames = append(columnNames, k)
		columnValues = append(columnValues, v)
		updateStatements = append(updateStatements, fmt.Sprintf("%s = $%d", k, index))
	}

	query := fmt.Sprintf(`INSERT INTO %s.promscale_instance_information(%s) VALUES (%s)
	ON CONFLICT (uuid) DO UPDATE SET %s`,
		schema.PromscaleCatalog,
		strings.Join(columnNames, ", "),
		strings.Join(indexes, ", "),
		strings.Join(updateStatements, ", "),
	)
```

But do NOTE: Metrics are registered individually by each module, meaning when the insert starts, some modules might not have registered their metrics. Hence, I will have to remove the `NOT NULL` constrain, even though this won't matter when update happens at `55 mins`.